### PR TITLE
Event driven architecture refactored.

### PR DESCRIPTION
**Pull request by legka**
1. Renamed Message to Event and Event to AbstractEvent
2. Generified Event and Handler to avoid casts in Handlers
3. Updated EventDispatcher to make unsafe configuration impossible (handler for wrong type)
4. Updated UML diagram accordingly

### CODE REVIEW

File: event-driven-architecture/src/main/java/com/iluwatar/eda/App.java [hunk 1]
```diff
@@ -53,12 +53,12 @@
   public static void main(String[] args) {
 
     EventDispatcher dispatcher = new EventDispatcher();
-    dispatcher.registerChannel(UserCreatedEvent.class, new UserCreatedEventHandler());
```
Comment (legka) [inline event-driven-architecture/src/main/java/com/iluwatar/eda/App.java]: Here you could actually have something like this:
```
dispatcher.registerChannel(UserCreatedEvent.class, new UserCreatedEventHandler());
dispatcher.registerChannel(UserUpdatedEvent.class, new UserUpdatedEventHandler());
```

The updated dispatcher will not allow this.

File: event-driven-architecture/src/main/java/com/iluwatar/eda/framework/EventDispatcher.java [hunk 2]
```diff
@@ -22,19 +22,16 @@
...
 
 /**
  * Handles the routing of {@link Event} messages to associated handlers.
  * A {@link HashMap} is used to store the association between events and their respective handlers.
- *
  */
 public class EventDispatcher {
 
-  private Map<Class<? extends Event>, Handler<?>> handlers;
+  private Map<Class<? extends Event>, Handler<? extends Event>> handlers;
```
Comment (legka) [inline event-driven-architecture/src/main/java/com/iluwatar/eda/framework/EventDispatcher.java]: This map declaration and the `registerHandler()` signature below, both ensure only 'type-safe' combination of event type and handler can be registered.

**Comment by npathai** [comment]
:+1: Good changes. Thanks for your contribution.

**Comment by legka** [comment]
My pleasure )

**Pull request merged**